### PR TITLE
Rfc/global naming conventions

### PR DESCRIPTION
**Pull request by dmoverton**
### Description

Add RFC containing proposal for global naming conventions.

### CODE REVIEW

File: rfcs/global-naming-conventions.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+However we still need to capitalise the first letter when it's used as the second word of a field name, e.g. `custom_name` of `users` -> field `deleteUsers`.
+1. Users can still override individual root fields names using `custom_root_fields`, e.g. if they really want to use `mySchemaUserAccount` instead of `mySchemaUserAccountsByPk`.
+1. I have only considered Postgres databases so far. Hopefully this proposal should carry over to other backends without problems.
+1. The [Apollo naming conventions](https://www.apollographql.com/docs/apollo-server/schema/schema/#naming-conventions) also mention that enum values should be in `ALL_CAPS`. Enum value names are currently beyond the scope of this RFC.
+
+### Questions
+1. Where should this global option go in the metadata and how should it be set? Or should it be a command line option / environment variable? - Global option in metadata
+1. Do we need an option per source or is a global option better? - global option
+1. Should it be `ByPk` or `ByPK`? `StddevPop` or `StdDevPop`?
```
Comment (elitan) [inline rfcs/global-naming-conventions.md:57]: `byPk` and `StdDevPop` is my suggestion.

File: rfcs/global-naming-conventions.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+However we still need to capitalise the first letter when it's used as the second word of a field name, e.g. `custom_name` of `users` -> field `deleteUsers`.
+1. Users can still override individual root fields names using `custom_root_fields`, e.g. if they really want to use `mySchemaUserAccount` instead of `mySchemaUserAccountsByPk`.
+1. I have only considered Postgres databases so far. Hopefully this proposal should carry over to other backends without problems.
+1. The [Apollo naming conventions](https://www.apollographql.com/docs/apollo-server/schema/schema/#naming-conventions) also mention that enum values should be in `ALL_CAPS`. Enum value names are currently beyond the scope of this RFC.
+
+### Questions
+1. Where should this global option go in the metadata and how should it be set? Or should it be a command line option / environment variable? - Global option in metadata
+1. Do we need an option per source or is a global option better? - global option
+1. Should it be `ByPk` or `ByPK`? `StddevPop` or `StdDevPop`? 
+1. What to do about comparison and boolean operators that start with `_`, e.g. `_eq`, `_and`, `_is_null`?
```
Comment (elitan) [inline rfcs/global-naming-conventions.md:58]: `_isNull` would be great 👍

File: rfcs/global-naming-conventions.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+### Notes
+
+1. If the user specifies a `custom_name` for a table, we should use it as is, without trying to convert to `camelCase`. The user can use this to override the automated case conversion if there are edge cases where it doesn't do what they want.
+However we still need to capitalise the first letter when it's used as the second word of a field name, e.g. `custom_name` of `users` -> field `deleteUsers`.
+1. Users can still override individual root fields names using `custom_root_fields`, e.g. if they really want to use `mySchemaUserAccount` instead of `mySchemaUserAccountsByPk`.
+1. I have only considered Postgres databases so far. Hopefully this proposal should carry over to other backends without problems.
+1. The [Apollo naming conventions](https://www.apollographql.com/docs/apollo-server/schema/schema/#naming-conventions) also mention that enum values should be in `ALL_CAPS`. Enum value names are currently beyond the scope of this RFC.
+
+### Questions
+1. Where should this global option go in the metadata and how should it be set? Or should it be a command line option / environment variable? - Global option in metadata
```
Comment (elitan) [inline rfcs/global-naming-conventions.md:55]: I think it would be best to have this as a global environment variable and metadata configuration can override using `custom_root_fields`.

File: rfcs/global-naming-conventions.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+| `delete_my_schema_user_accounts` | `deleteMySchemaUserAccounts` |
+| `delete_my_schema_user_accounts_by_pk` | `deleteMySchemaUserAccountsByPk` |
+| `insert_my_schema_user_accounts` | `insertMySchemaUserAccounts` |
+| `insert_my_schema_user_accounts_one` | `insertMySchemaUserAccountsOne` |
+| `update_my_schema_user_accounts` | `updateMySchemaUserAccounts` |
+| `update_my_schema_user_accounts_by_pk` | `updateMySchemaUserAccountsByPk` |
+
+We would also convert all column names to `camelCase`.
+Similarly, all type names would use `PascalCase`, e.g. `my_schema_user_accounts_aggregate_fields` would become `MySchemaUserAccountsAggregateFields`.
+We would also modify other generated names e.g. `distinct_on` -> `distinctOn` and `stdev_pop` -> `StddevPop`.
```
Comment (jcarrus) [inline rfcs/global-naming-conventions.md:44]:
```suggestion
We would also modify other generated names e.g. `distinct_on` -> `distinctOn` and `stddev_pop` -> `stddevPop`.
```

File: rfcs/global-naming-conventions.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+However we still need to capitalise the first letter when it's used as the second word of a field name, e.g. `custom_name` of `users` -> field `deleteUsers`.
+1. Users can still override individual root fields names using `custom_root_fields`, e.g. if they really want to use `mySchemaUserAccount` instead of `mySchemaUserAccountsByPk`.
+1. I have only considered Postgres databases so far. Hopefully this proposal should carry over to other backends without problems.
+1. The [Apollo naming conventions](https://www.apollographql.com/docs/apollo-server/schema/schema/#naming-conventions) also mention that enum values should be in `ALL_CAPS`. Enum value names are currently beyond the scope of this RFC.
+
+### Questions
+1. Where should this global option go in the metadata and how should it be set? Or should it be a command line option / environment variable? - Global option in metadata
+1. Do we need an option per source or is a global option better? - global option
```
Comment (tirumaraiselvan) [inline rfcs/global-naming-conventions.md:56]: I think having this per source might be better because different databases may have different conventions and the API consumer would like to follow those conventions.

File: rfcs/global-naming-conventions.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+### Notes
+
+1. If the user specifies a `custom_name` for a table, we should use it as is, without trying to convert to `camelCase`. The user can use this to override the automated case conversion if there are edge cases where it doesn't do what they want.
+However we still need to capitalise the first letter when it's used as the second word of a field name, e.g. `custom_name` of `users` -> field `deleteUsers`.
+1. Users can still override individual root fields names using `custom_root_fields`, e.g. if they really want to use `mySchemaUserAccount` instead of `mySchemaUserAccountsByPk`.
+1. I have only considered Postgres databases so far. Hopefully this proposal should carry over to other backends without problems.
+1. The [Apollo naming conventions](https://www.apollographql.com/docs/apollo-server/schema/schema/#naming-conventions) also mention that enum values should be in `ALL_CAPS`. Enum value names are currently beyond the scope of this RFC.
+
+### Questions
+1. Where should this global option go in the metadata and how should it be set? Or should it be a command line option / environment variable? - Global option in metadata
```
Comment (tirumaraiselvan) [inline rfcs/global-naming-conventions.md:55]: We can have a field called `naming_convention` in the `customization` field of each source in the metadata.

File: rfcs/global-naming-conventions.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+| `update_my_schema_user_accounts` | `updateMySchemaUserAccounts` |
+| `update_my_schema_user_accounts_by_pk` | `updateMySchemaUserAccountsByPk` |
+
+We would also convert all column names to `camelCase`.
+Similarly, all type names would use `PascalCase`, e.g. `my_schema_user_accounts_aggregate_fields` would become `MySchemaUserAccountsAggregateFields`.
+We would also modify other generated names e.g. `distinct_on` -> `distinctOn` and `stddev_pop` -> `stddevPop`.
+
+### Notes
+
+1. If the user specifies a `custom_name` for a table, we should use it as is, without trying to convert to `camelCase`. The user can use this to override the automated case conversion if there are edge cases where it doesn't do what they want.
```
Comment (tirumaraiselvan) [inline rfcs/global-naming-conventions.md:48]: Currently, if the table has non-graphql compliant characters then it's discarded from the schema. We should continue to do that with the naming convention capability as well.

**Comment by xiaoyu-tamu** [comment]
can we support PascalCase `type names` only mode? 

e.g. `my_schema_user_accounts_aggregate_fields` would become `MySchemaUserAccountsAggregateFields`.

and custom root field `mySchemaUserAccount` instead of `mySchemaUserAccountsByPk`.

**Comment by tirumaraiselvan** [comment]
> can we support PascalCase `type names` only mode?
> 
> e.g. `my_schema_user_accounts_aggregate_fields` would become `MySchemaUserAccountsAggregateFields`.

Any specific reason you want field names to be snake case and type names to be pascal case? Is this a known convention?

**Comment by xiaoyu-tamu** [comment]
> > can we support PascalCase `type names` only mode?
> > e.g. `my_schema_user_accounts_aggregate_fields` would become `MySchemaUserAccountsAggregateFields`.
> 
> Any specific reason you want field names to be snake case and type names to be pascal case? Is this a known convention?

My bad, I didn't say it clearly. 

I want the field name and enum name to be PascalCase and let Hasura generate it. 

e.g. `MySchemaUserAccountsAggregateFields` and `MyEnumType`

For type name, I want to override and use my own naming strategy with the `custom_root_fields` settings to avoid break changes for my API. 

`mySchemaUserAccount` instead of `mySchemaUserAccountByPk`

To be short, I hope `custom_root_fields` has highest priority.

**Comment by tirumaraiselvan** [comment]
> To be short, I hope custom_root_fields has highest priority.

Yeah! The priority order is going to be naming convention < custom name < custom root fields.

File: rfcs/global-naming-conventions.md (hunk 5 above)
Comment (beaussan) [reply rfcs/global-naming-conventions.md:56]: Having a per source might give users more fine control at the cost of consistency, especially if the operators also include the naming convention, so that is also something to consider.

File: rfcs/global-naming-conventions.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+However we still need to capitalise the first letter when it's used as the second word of a field name, e.g. `custom_name` of `users` -> field `deleteUsers`.
+1. Users can still override individual root fields names using `custom_root_fields`, e.g. if they really want to use `mySchemaUserAccount` instead of `mySchemaUserAccountsByPk`.
+1. I have only considered Postgres databases so far. Hopefully this proposal should carry over to other backends without problems.
+1. The [Apollo naming conventions](https://www.apollographql.com/docs/apollo-server/schema/schema/#naming-conventions) also mention that enum values should be in `ALL_CAPS`. Enum value names are currently beyond the scope of this RFC.
```
Comment (tirumaraiselvan) [inline rfcs/global-naming-conventions.md:52]:
```suggestion
```

Let's include this in the current scope? Otherwise, it is a bit of an inconsistency.

Comment (kdawgwilk) [reply rfcs/global-naming-conventions.md:52]: Agreed 👍

File: rfcs/global-naming-conventions.md (hunk 2 above)
Comment (lemohammed) [reply rfcs/global-naming-conventions.md:58]: or simply remove the `_` as well, something like `and`, `eq`, `isNull`

**Comment by tirumaraiselvan** [comment]
Experimental support for this has been added in `v2.8.0-beta.1` . Please see the docs for usage and current limitations (postgres only): https://hasura.io/docs/latest/graphql/core/databases/postgres/schema/naming-convention/

**Comment by paritosh-08** [comment]
`graphql-default` naming convention raised a question on how hasura adds prefix and suffix to the field names. I have added an issue with more details [here](https://github.com/hasura/graphql-engine/issues/8544).

TLDR: If we add prefix `foo` to a field name `myTableByPk`, should the name be `fooMyTableByPk` or should it be `foomyTableByPk`? If former, should we do a similar thing (like adding `_`) with `hasura-default` naming convention?
